content/non-blocking-communication-pt1.rst

```diff
@@ -1,33 +1,307 @@
+-----------------------------------------------------
+
+.. challenge:: 1.1 Observe a deadlock
+
+   1. Download the :download:`source code
+      <code/non-blocking-communication-deadlock.c>`. Open
+      ``non-blocking-communication-deadlock.c`` and read through it. Compile
+      with::
+
+        mpicc -g -Wall -std=c11 non-blocking-communication-deadlock.c -o non-blocking-communication-deadlock
+
+   2. When you have the code compiling, try to run with::
+
+        mpiexec -np 2 ./non-blocking-communication-deadlock
+
+   3. The communication may block. If it does, you will have to kill
+      the process to continue, e.g. with ``Ctrl-C``. If it doesn't,
+      follow the first challenge to use a call to |term-MPI_Ssend|
+      to make it block.
+
+   4. Try to fix the code so that one process sends before receiving
+      and the other process does the opposite. Now it will work even
+      if the runtime chooses to implement |term-MPI_Send| like
+      |term-MPI_Ssend|.
+
+.. solution::
+
+   * One correct approach is::
+
+        /* Do sends and receives in the opposite order on the two ranks */
+        if (rank == 0)
+        {
+            int send_up_tag = 0, send_down_tag = 1;
+            /* Send the border data */
+            int destination_rank = size-rank-1;
+            MPI_Ssend(working_data_set[1], 8, MPI_INT, destination_rank, send_up_tag, comm);
+            MPI_Ssend(working_data_set[4], 8, MPI_INT, destination_rank, send_down_tag, comm);
+
+            /* Receive the halo data */
+            int source_rank = size-rank-1;
+            MPI_Recv(working_data_set[5], 8, MPI_INT, source_rank, send_up_tag, comm, MPI_STATUS_IGNORE);
+            MPI_Recv(working_data_set[0], 8, MPI_INT, source_rank, send_down_tag, comm, MPI_STATUS_IGNORE);
+        }
+        else
+        {
+            int send_up_tag = 0, send_down_tag = 1;
+            /* Receive the halo data */
+            int source_rank = size-rank-1;
+            MPI_Recv(working_data_set[5], 8, MPI_INT, source_rank, send_up_tag, comm, MPI_STATUS_IGNORE);
+            MPI_Recv(working_data_set[0], 8, MPI_INT, source_rank, send_down_tag, comm, MPI_STATUS_IGNORE);
+
+            /* Send the border data */
+            int destination_rank = size-rank-1;
+            MPI_Ssend(working_data_set[1], 8, MPI_INT, destination_rank, send_up_tag, comm);
+            MPI_Ssend(working_data_set[4], 8, MPI_INT, destination_rank, send_down_tag, comm);
+        }
+
+   * There are other approaches that work correctly. Is yours better
+     or worse than this one? Why?
+   * Download a :download:`working solution <code/non-blocking-communication-deadlock-solution.c>`
+
+     
 See also
 --------
 
 
-* The lecture covering MPI+OpenMP from EPCC is available `here <http://www.archer.ac.uk/training/course-material/2020/01/advMPI-imperial/Slides/L06-MPIandOpenMP.pdf>`_
+* TODO
+* TODO
 * Chapter 6 of the **Using Advanced MPI** by William Gropp *et al.* :cite:`Gropp2014-dz`
```

Comment:
The "See also" is also not pertinent to this episode.

Reply:
0k